gym/envs/mujoco/mujoco_env.py

```diff
@@ -30,6 +44,57 @@ def __init__(self, model_path, frame_skip):
             'video.frames_per_second' : int(np.round(1.0 / self.dt))
         }
 
+        self.init_qpos = self.model.data.qpos.ravel().copy()
+        self.init_qvel = self.model.data.qvel.ravel().copy()
+        observation, _reward, done, _info = self.step(np.zeros(self.model.nu))
+        assert not done
+        self.obs_dim = observation.size
+
+        bounds = self.model.actuator_ctrlrange.copy()
+        low = bounds[:, 0]
+        high = bounds[:, 1]
+        self.action_space = spaces.Box(low, high)
+
+        high = np.inf*np.ones(self.obs_dim)
+        low = -high
+        self.observation_space = spaces.Box(low, high)        
+
+    # methods to override:
+    # ----------------------------
+
+    def reset_model(self):
+        """
+        Reset the robot degrees of freedom (qpos and qvel).
+        Implement this in each subclass.
+        """
+        raise NotImplementedError
+
+    def viewer_setup(self):
+        """
+        This method is called when the viewer is initialized and after every reset
+        Optionally implement this method, if you need to tinker with camera position
+        and so forth.
+        """
+        pass
+
+    # -----------------------------
+
+    def _reset(self):
+        ob = self.reset_model()
+        # mjlib.mj_resetData(self.model.ptr, self.data.ptr)
```

Comment:
Is this supposed to be commented out? Also, I can't say for sure, but I would think that `mj_resetData` should go before `reset_model`

Also, take a look at http://www.mujoco.org/book/file/simulate.cpp
It contains the following code for resetting a simulation:

```
case GLFW_KEY_BACKSPACE:            // reset
        mj_resetData(m, d);
        mj_forward(m, d);
        break;
```

Maybe the canonical way to reset is to call these two functions in sequence.